### PR TITLE
fix regression with non-object iterables (functions) being renderable

### LINKED ISSUE #11396: A function that is also an iterable is treated inconsistently
**Do you want to request a *feature* or report a *bug*?**
*bug* (I think)

**What is the current behavior?**
Here's a [codesandbox](https://codesandbox.io/s/0ml3roomr0). Or just plop this in an `index.html`:

```html
<body>
<div id="root"></div>
<script src="https://unpkg.com/react@16.0.0/umd/react.development.js"></script>
<script src="https://unpkg.com/react-dom@16.0.0/umd/react-dom.development.js"></script>
<script src="https://unpkg.com/babel-standalone@6.26.0/babel.js"></script>
<script type="text/babel">
function makeRenderable(fn, valueGetter) {
  const ITERATOR_SYMBOL =
    (typeof Symbol === 'function' && Symbol.iterator) || '@@iterator'
  function iterator() {
    let timesCalled = 0
    return {
      next() {
        const done = timesCalled++ > 0
        return {done, value: done ? undefined : valueGetter()}
      },
    }
  }
  fn[ITERATOR_SYMBOL] = iterator
  return fn
}

const renderableFunction = makeRenderable(function(a, b) {
  // kinda irrelevant what goes on in here...
  return a + b
}, () => 'I am the rendered version of the function')

function Working() {
  return (
    <div>
      This works: <div>{renderableFunction}</div>
    </div>
  )
}

function NotWorking() {
  return <div>This does not work: {renderableFunction}</div>
}

ReactDOM.render(
  <div>
    <Working />
    <hr />
    <NotWorking />
  </div>,
  document.getElementById('root'),
)
</script>
</body>
```

**What is the expected behavior?**

I expect that things will render the same regardless of whether they're an only child or a member of an array of children.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

Yes, [it worked in React@15](https://codesandbox.io/s/8zlpvjzjy9).

**Suggested Solution**

I'm fairly confident that the issue is in the array of children case, the [`getIteratorFn` call](https://github.com/facebook/react/blob/2c0a8fb99e945315c9dce7a15c85775d51f5755d/packages/react-reconciler/src/ReactChildFiber.js#L578-L586) is within a `typeof newChild === 'object' && newChild !== null` if statement, whereas the original [reconcileChildFibers](https://github.com/facebook/react/blob/2c0a8fb99e945315c9dce7a15c85775d51f5755d/packages/react-reconciler/src/ReactChildFiber.js#L1358) puts the [`getIteratorFn` call outside it](https://github.com/facebook/react/blob/2c0a8fb99e945315c9dce7a15c85775d51f5755d/packages/react-reconciler/src/ReactChildFiber.js#L1434-L1441).

If I'm not mistaken, a simple fix would be to move [this curly brace](https://github.com/facebook/react/blob/2c0a8fb99e945315c9dce7a15c85775d51f5755d/packages/react-reconciler/src/ReactChildFiber.js#L589) a few lines up :wink:

I'm happy to do so complete with a test given a little direction (like, where the test should go). Thanks!

### PR DESCRIPTION
**Pull request by kentcdodds**
- [x] Fork [the repository](https://github.com/facebook/react) and create your branch from `master`.
- [x] Run `yarn` in the repository root.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x]  Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
- [x]  Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
- [x]  Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
- [x]  Run the [Flow](https://flowtype.org/) typechecks (`yarn flow`).
- [x]  If you haven't already, complete the CLA.

This closes #11396

I noticed a few other spots where there would be issues with non-object iterables could cause problems and added checks for those as well.

### CODE REVIEW

File: packages/react-reconciler/src/ReactChildFiber.js [hunk 1]
```diff
@@ -794,6 +819,17 @@ function ChildReconciler(shouldTrackSideEffects) {
       throwOnInvalidObjectType(returnFiber, newChild);
     }
 
+    if (getIteratorFn(newChild)) {
```
[the remark this answers is not in the document]
Comment (gaearon) [reply packages/react-reconciler/src/ReactChildFiber.js]: Update inside a keyed array?

**Comment by kentcdodds** [comment]
I'll give it some time this afternoon 👍 Thanks!

**Comment by kentcdodds** [comment]
Really struggling to figure out how to get into this code path. Everything I try will exit here:

https://github.com/facebook/react/blob/913a125ad51777ffec91eedefa03c534a7590395/packages/react-reconciler/src/ReactChildFiber.js#L912-L916

Or here:

https://github.com/facebook/react/blob/913a125ad51777ffec91eedefa03c534a7590395/packages/react-reconciler/src/ReactChildFiber.js#L918-L940

I've tried a bunch of stuff. Here's my latest attempt:

```javascript
const getItems = () => [
  makeRenderableFunction('f1'),
  makeRenderableFunction('f2'),
  'non-fn',
];

class Parent extends React.Component {
  constructor() {
    super();
    this.state = {items: ['a', 'b', makeRenderableFunction('f3'), 'c', 'd']};
    this.setItems = items => this.setState({items});
  }
  render() {
    return (
      <div>
        {this.state.items.map((item, i) => <span key={i}>{item}</span>)}
      </div>
    );
  }
}

const comp = ReactTestUtils.renderIntoDocument(<Parent />);
comp.setItems(getItems());
```

Any tips appreciated :)

**Comment by gaearon** [comment]
You can throw in that code path, run all tests, and thus find the simplest (now failing) test the exercises this path. Then reverse engineer what it's going.

**Comment by kentcdodds** [comment]
Good idea. Thanks :+1:

**Comment by kentcdodds** [comment]
That definitely helped @gaearon! Thanks. I've updated the PR to include a test for that case as well.

**Comment by gaearon** [comment]
We've decided we don't want to support this.
It's a very very niche use case, and it's not worth it doing those extra checks.
So instead we'd like to take a PR that disables the (accidental) support completely.

**Pull request closed without merging**